src/ServicePulse.Host/app/modules/monitoring/js/endpoint_details.controller.js

```diff
@@ -34,20 +35,62 @@
             updateUI();
         };
 
-        function mergeIn(destination, source) {
+        function mergeIn(destination, source, propertiesToSkip) {
             for (var propName in source) {
                 if (source.hasOwnProperty(propName)) {
-                    destination[propName] = source[propName];
+                    if(!propertiesToSkip || !propertiesToSkip.includes(propName)) {
+                        destination[propName] = source[propName];
+                    }
                 }
             }
         }
 
-        $scope.buildUrl = function (selectedPeriodValue, showInstancesBreakdownTab) {
-            return `#/monitoring/endpoint/${$scope.endpointName}/${$scope.sourceIndex}?historyPeriod=${selectedPeriodValue}&tab=${$scope.showInstancesBreakdown ? 'instancesBreakdown' : 'messageTypeBreakdown'}`;
+        $scope.buildUrl = function (selectedPeriodValue, showInstacesBreakdown, breakdownPageNo) {
+
+            var breakdownTabName = showInstacesBreakdown ? 'instancesBreakdown' : 'messageTypeBreakdown';
+
+            return `#/monitoring/endpoint/${$scope.endpointName}/${$scope.sourceIndex}?historyPeriod=${selectedPeriodValue}&tab=${breakdownTabName}&pageNo=${breakdownPageNo}`;
+        };
+
+        $scope.updateUrl = function () {
+
+            var updatedUrl = $scope.buildUrl($scope.selectedPeriod.value, $scope.showInstancesBreakdown, $scope.endpoint.messageTypesPage);
+
+            window.location.hash = updatedUrl;
         };
 
         $scope.showInstancesBreakdownTab = function(isVisible) {
             $scope.showInstancesBreakdown = isVisible;
+
+            $scope.endpoint.messageTypesForceReload();
+        };
+
+        $scope.endpoint = {
+            messageTypesPage: !$scope.showInstancesBreakdown ? $routeParams.pageNo : 1,
+            messageTypesTotalItems: 0,
+            messageTypesItemsPerPage: 10,
+            messageTypesAvailable: false,
+            messageTypesUpdatedSet: [],
+            messageTypesForceReload: function () {
```

Comment:
Not a fan of this name. If it's a Force, where is the normal non-force reload? What about simply `reloadMessageTypes`?

Reply:
how about 'refreshMessageTypes`?